### ISSUE TITLE
stopgap idb replacement broke TS compilation under node due to usage of DOMStringList in @firebase/util 1.5.0

### ISSUE THREAD

**Issue by Ranguna** [firebase/firebase-js-sdk#6085]
### [REQUIRED] Describe your environment

  * Operating System version: Linux
  * Browser version: none, I'm using nodejs
  * Firebase SDK version: latest
  * Firebase Product: firebase util

#### Steps to reproduce:
Install @firebase/util and compile under node environment.

To reproduce with `firebase-admin` (which has a dependency chain to `@firebase/util`), you can checkout this repo: https://github.com/Ranguna/reproducing-firebase-js-sdk-6085

#### Relevant Code:
https://github.com/firebase/firebase-js-sdk/commit/2d672cead167187cb714cd89b638c0884ba58f03#diff-b8c22a8c7d8e12d84f62d2ee45dffda84efae4b8fa6588d1aac8f0d49909fd57R33
The above line introduced the usage of `DOMStringList` which is not available in node. This breaks ts compilations for `firebase-admin` through this dependency chain `firebase-admin -> @firebase/database-compat -> @firebase/util`.

I am now stuck with `@firebase/util` < 1.5.0.

Typescript compilation result:
```
> tsc

node_modules/@firebase/util/dist/src/indexeddb.d.ts:19:23 - error TS2304: Cannot find name 'DOMStringList'.

19     objectStoreNames: DOMStringList;
                         ~~~~~~~~~~~~~

node_modules/@firebase/util/dist/src/indexeddb.d.ts:20:22 - error TS2304: Cannot find name 'IDBDatabase'.

20     constructor(_db: IDBDatabase);
                        ~~~~~~~~~~~

node_modules/@firebase/util/dist/src/indexeddb.d.ts:21:55 - error TS2304: Cannot find name 'IDBTransactionMode'.

21     transaction(storeNames: string[] | string, mode?: IDBTransactionMode): TransactionWrapper;
                                                         ~~~~~~~~~~~~~~~~~~

node_modules/@firebase/util/dist/src/indexeddb.d.ts:22:52 - error TS2304: Cannot find name 'IDBObjectStoreParameters'.

22     createObjectStore(storeName: string, options?: IDBObjectStoreParameters): ObjectStoreWrapper;
                                                      ~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@firebase/util/dist/src/indexeddb.d.ts:28:31 - error TS2304: Cannot find name 'IDBTransaction'.

28     constructor(_transaction: IDBTransaction);
                                 ~~~~~~~~~~~~~~

node_modules/@firebase/util/dist/src/indexeddb.d.ts:33:25 - error TS2304: Cannot find name 'IDBObjectStore'.

33     constructor(_store: IDBObjectStore);
                           ~~~~~~~~~~~~~~

node_modules/@firebase/util/dist/src/indexeddb.d.ts:35:57 - error TS2304: Cannot find name 'IDBIndexParameters'.

35     createIndex(name: string, keypath: string, options: IDBIndexParameters): IndexWrapper;
                                                           ~~~~~~~~~~~~~~~~~~

node_modules/@firebase/util/dist/src/indexeddb.d.ts:43:25 - error TS2304: Cannot find name 'IDBIndex'.

43     constructor(_index: IDBIndex);
                           ~~~~~~~~


Found 8 errors in the same file, starting at: node_modules/@firebase/util/dist/src/indexeddb.d.ts:19
```

**Comment by Ranguna** [comment]
For anyone coming here with the same problem, you can add the following to you `package.json` as a temporary solutions:
```
"overrides": {
    "@firebase/util@>1.4.3": "1.4.3"
  }
```
(works under npm 8.3+)

**Comment by hsubox76** [comment]
This workaround may work for admin users since you won't have any of the idb dependent packages that were changed (app, analytics, app-check, performance, remote-config, messaging), but for anyone using `firebase` directly, I don't recommend downgrading `@firebase/util` by itself because those other packages previously listed were also updated to import some new methods added to util, and those methods won't be there in the old util version and may break the build. For any `firebase` users having this problem (who are not `firebase-admin` users) I'd recommend downgrading the entire `firebase` package as a whole.

In any case I will try to get out a fix ASAP.

**Comment by hsubox76** [comment]
Sorry, I'm not really used to TS in Node and I'm trying to reproduce and I'm wondering if you can help me figure out how to do so. I've made a repo that has the deps:
```
    "firebase-admin": "^10.0.2",
    "ts-node": "^10.7.0",
    "typescript": "^4.6.2"
```
And I've made a simple `index.ts`:
```
import { initializeApp, applicationDefault } from "firebase-admin/app";
import { getDatabase } from 'firebase-admin/database';

initializeApp({
  credential: applicationDefault(),
  databaseURL: /** db url here **/
});
getDatabase();
```
And I've run that file with both `tsc index.ts` and `ts-node index.ts` and haven't gotten any errors. What do I need to change to reproduce this?

**Comment by Ranguna** [comment]
@hsubox76 Thanks for the headsup about the workaround!

I'm extremely sorry, I should've added better steps on how to reproduce this. To make it easier, I created this repo: https://github.com/Ranguna/reproducing-firebase-js-sdk-6085

Run `npm i` followed by `npm run build` and you should see the compilation errors pop up.

I'll update the OP with this info.

**Comment by Ranguna** [comment]
Also, another workaround is to set this in your tsconfig:
```
{
  "compilerOptions": {
    ...,
    "skipLibCheck": true
  }
}
```

But that's also extremely risky because it's going to skip type checks for all packages, not just firebase.

**Comment by Ranguna** [comment]
> > Also, another workaround is to set this in your tsconfig:
> > 
> > ```
> > {
> >   "compilerOptions": {
> >     ...,
> >     "skipLibCheck": true
> >   }
> > }
> > ```
> > 
> > But that's also extremely risky because it's going to skip type checks for all packages, not just firebase.
> > 
> > EDIT: This workaround will possibly break your app if you are running under node, since the introduced types are not available in a node environment. I've seen similar issues before that have worked in node, so there's a chance it's ok. Make sure you test your app thoroughly if you decide to go with this route.
> 
> Thank you - this worked for us.

Just FYI, "worked" is relative here.

You just lost maybe 50% usefulness of typescript by toggling that setting. Since probably 99% of your total bundled code is located inside node_modules and maybe half of it is written in typescript (wishful thinking, I know); and you have now disabled all type checks around this.

Like I said on my parent comment, make sure you test you code thoroughly; and no, unit tests are probably not enough. Just a friendly heads-up.

Also, don't forget to toggle it back once this is fixed.

**Comment by TacB0sS** [comment]
> This workaround may work for admin users since you won't have any of the idb dependent packages that were changed (app, analytics, app-check, performance, remote-config, messaging), but for anyone using `firebase` directly, I don't recommend downgrading `@firebase/util` by itself because those other packages previously listed were also updated to import some new methods added to util, and those methods won't be there in the old util version and may break the build. For any `firebase` users having this problem (who are not `firebase-admin` users) I'd recommend downgrading the entire `firebase` package as a whole.
> 
> In any case I will try to get out a fix ASAP.

True.. I cannot compile my project due to this issue, and I am still trying to figure which version of firebase would allow me to continue working

Any ETA on this issue?

**Comment by hsubox76** [comment]
Forgot to link the issue in the PR but this PR should fix it and the release should go out today https://github.com/firebase/firebase-js-sdk/pull/6088

**Comment by hsubox76** [comment]
It's been released in [9.6.10](https://firebase.google.com/support/release-notes/js#version_9610_-_march_24_2022), please let me know if there are still any issues.

**Comment by Ranguna** [comment]
Confirmed, I can no longer reproduce the issue.

Thank you very much for the fix!
